service_state.go

```diff
@@ -0,0 +1,123 @@
+package baps3
+
+// The part of the BAPS3 connector code responsible for updating the
+// internal state.
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+)
+
+// ServiceState is the struct of all known state for a BAPS3 service.
+// TODO(CaptainHayashi): possibly segregate more by feature, so elements not
+// relevant to the current feature set aren't allocated?
+type ServiceState struct {
+	// Core
+	Features FeatureSet
+	State    string
+
+	// TimeReport
+	Time time.Duration
+
+	// FileLoad
+	File string
+}
+
+// InitServiceState creates a new, blank, ServiceState.
+func InitServiceState() (s *ServiceState) {
+	s = new(ServiceState)
+	s.Features = FeatureSet{}
+	s.State = "Ready"
+
+	return
+}
+
+// Update updates a ServiceState according to an incoming service response.
+func (s *ServiceState) Update(res Message) (err error) {
+	switch res.Word() {
+	case RsFeatures:
+		err = s.updateFeaturesFromMessage(res)
+	case RsFile:
+		err = s.updateFileFromMessage(res)
+	case RsState:
+		err = s.updateStateFromMessage(res)
+	case RsTime:
+		err = s.updateTimeFromMessage(res)
+	}
+
+	return
```

Comment:
I'm the one that prefers named returns :grinning:  It can sometimes save a bit of code etc (especially when you want to return the zero value in some cases, then you don't need to initialise the variable yourself).

But here you probably either want normal returns, or a `map[MessageWord]func(*ServiceState, Message)error`, because the named return doesn't really give you any advantages.